File: agents/orchestrator.py
"""Orchestrator Agent - Coordinates multi-agent task execution."""

from __future__ import annotations

from dataclasses import dataclass, field

from agents.implementation_agent import ImplementationAgent
from agents.security_agent import SecurityAgent
from agents.testing_agent import TestingAgent


@dataclass
class SubTask:
    name: str
    agent_type: str
    description: str
    dependencies: list[str] = field(default_factory=list)
    result: str | None = None


class OrchestratorAgent:
    """Decomposes tasks into subtasks and delegates to specialist agents."""

    def __init__(self, api_key: str | None = None):
        self.implementation = ImplementationAgent(api_key=api_key)
        self.security = SecurityAgent()
        self.testing = TestingAgent(api_key=api_key)
        self._agents = {
            "implementation": self.implementation,
            "security": self.security,
            "testing": self.testing,
        }

    def decompose_task(self, task: str) -> list[SubTask]:
        """Split a high-level task into ordered subtasks for specialist agents."""
        return [
            SubTask(
                name="implement",
                agent_type="implementation",
                description=f"Implement: {task}",
            ),
            SubTask(
                name="security_review",
                agent_type="security",
                description="Review implementation for security issues",
                dependencies=["implement"],
            ),
            SubTask(
                name="generate_tests",
                agent_type="testing",
                description="Generate unit tests for the implementation",
                dependencies=["implement"],
            ),
        ]

    def run(self, task: str) -> dict:
        """Execute a full multi-agent workflow for the given task."""
        subtasks = self.decompose_task(task)
        results: dict[str, str] = {}

        for subtask in subtasks:
            agent = self._agents[subtask.agent_type]

            if subtask.agent_type == "implementation":
                subtask.result = agent.run(subtask.description)
            elif subtask.agent_type == "security":
                code = results.get("implement", "")
                review = agent.run(code)
                subtask.result = str(review)
            elif subtask.agent_type == "testing":
                code = results.get("implement", "")
                subtask.result = agent.run(code)

            results[subtask.name] = subtask.result or ""

        return {
            "task": task,
            "code": results.get("implement", ""),
            "security_review": results.get("security_review", ""),
            "tests": results.get("generate_tests", ""),
        }

File: src/orchestrator/orchestrator.ts
import { TaskDAG } from "./dag";
import { AgentRole, ExecutionPlan, OrchestratorResult, Task, TaskResult } from "./types";
import { BaseAgent } from "../agents/base-agent";
import { ImplementationAgent } from "../agents/implementation-agent";
import { TestingAgent } from "../agents/testing-agent";
import { DocumentationAgent } from "../agents/documentation-agent";
import { SecurityAgent } from "../agents/security-agent";
import { createChildLogger } from "../utils/logger";

export class Orchestrator {
  private dag: TaskDAG;
  private agents: Map<AgentRole, BaseAgent>;
  private logger = createChildLogger("orchestrator");

  constructor() {
    this.dag = new TaskDAG();
    this.agents = new Map<AgentRole, BaseAgent>([
      ["implementation", new ImplementationAgent()],
      ["testing", new TestingAgent()],
      ["documentation", new DocumentationAgent()],
      ["security", new SecurityAgent()],
    ]);
  }

  decompose(name: string, description: string): TaskDAG {
    this.dag = new TaskDAG();
    let taskCounter = 0;
    const makeId = () => `task-${++taskCounter}`;

    // Phase 1: Implementation (no dependencies)
    const implId = makeId();
    this.dag.addTask({
      id: implId,
      name: `Implement: ${name}`,
      description,
      assignedTo: "implementation",
      status: "pending",
      dependencies: [],
    });

    // Phase 2: Testing + Documentation (depend on implementation)
    const testId = makeId();
    this.dag.addTask({
      id: testId,
      name: `Test: ${name}`,
      description: `Generate and run tests for ${description}`,
      assignedTo: "testing",
      status: "pending",
      dependencies: [implId],
    });

    const docId = makeId();
    this.dag.addTask({
      id: docId,
      name: `Document: ${name}`,
      description: `Generate documentation for ${description}`,
      assignedTo: "documentation",
      status: "pending",
      dependencies: [implId],
    });

    // Phase 3: Security (depends on implementation)
    const secId = makeId();
    this.dag.addTask({
      id: secId,
      name: `Security Review: ${name}`,
      description: `Security scan for ${description}`,
      assignedTo: "security",
      status: "pending",
      dependencies: [implId],
    });

    this.logger.info(`Decomposed "${name}" into ${taskCounter} tasks`);
    return this.dag;
  }

  plan(): ExecutionPlan {
    return this.dag.buildExecutionPlan();
  }

  async execute(): Promise<OrchestratorResult> {
    const executionPlan = this.plan();
    const start = Date.now();
    const phaseResults: { phase: number; results: TaskResult[] }[] = [];

    this.logger.info(
      `Executing plan: ${executionPlan.totalTasks} tasks in ${executionPlan.phases.length} phases`,
    );

    for (const phase of executionPlan.phases) {
      this.logger.info(
        `Phase ${phase.phaseNumber}: ${phase.tasks.length} tasks (parallel: ${phase.parallel})`,
      );

      const results = await this.executePhase(phase.tasks, phase.parallel);
      phaseResults.push({ phase: phase.phaseNumber, results });

      const allSucceeded = results.every((r) => r.success);
      if (!allSucceeded) {
        this.logger.warn(`Phase ${phase.phaseNumber} had failures, continuing...`);
      }
    }

    const totalDuration = Date.now() - start;
    const allResults = phaseResults.flatMap((p) => p.results);
    const successCount = allResults.filter((r) => r.success).length;

    return {
      success: successCount === allResults.length,
      phases: phaseResults,
      totalDuration,
      summary: `${successCount}/${allResults.length} tasks succeeded in ${totalDuration}ms`,
    };
  }

  private async executePhase(tasks: Task[], parallel: boolean): Promise<TaskResult[]> {
    if (parallel) {
      return Promise.all(tasks.map((task) => this.executeTask(task)));
    }
    const results: TaskResult[] = [];
    for (const task of tasks) {
      results.push(await this.executeTask(task));
    }
    return results;
  }

  private async executeTask(task: Task): Promise<TaskResult> {
    const agent = task.assignedTo ? this.agents.get(task.assignedTo) : undefined;
    if (!agent) {
      return {
        success: false,
        output: "",
        artifacts: [],
        duration: 0,
        errors: [`No agent assigned for role: ${task.assignedTo}`],
      };
    }

    task.status = "running";
    const result = await agent.run(task);
    task.status = result.success ? "completed" : "failed";
    task.result = result;
    return result;
  }

  getDAG(): TaskDAG {
    return this.dag;
  }
}
